src/USER-PHONON/dynamical_matrix.cpp

```diff
@@ -0,0 +1,511 @@
+#else
+    fp = popen(gzip,"w");
+#endif
+#else
+        error->one(FLERR,"Cannot open gzipped file");
+#endif
+    } else if (binaryflag) {
+        fp = fopen(filename,"wb");
+    } else {
+        fp = fopen(filename,"w");
+    }
+
+    if (fp == NULL) error->one(FLERR,"Cannot open dump file");
+
+    file_opened = 1;
+}
+
+/* ----------------------------------------------------------------------
+   create dynamical matrix
+------------------------------------------------------------------------- */
+
+void DynamicalMatrix::calculateMatrix(char *arg)
+{
+    int nlocal = atom->nlocal;
+    int plocal;
+    int id;
+    int group_flag = 0;
+    int *mask = atom->mask;
+    int *type = atom->type;
+    tagint *aid = atom->tag; //atom id
+    double dyn_element[nlocal][3]; //first index is nlocal (important)
+    double moaoi; //mass of atom of interest
+    double imass;
+    double del = force->numeric(FLERR, arg);
+    double *m = atom->mass;
+    double **x = atom->x;
+    double **f = atom->f;
+
+
+    //initialize dynmat to all zeros
+    for (int i=0; i < dynlen; i++)
+        for (int j=0; j < dynlen; j++)
+            dynmat[i][j] = 0.;
+
+
+    if (strstr(arg,"v_") == arg) error->all(FLERR,"Variable for dynamical_matrix is not supported");
+
+    energy_force(0);
+
+
+    if (comm->me == 0 && screen) fprintf(screen,"Calculating Dynamical Matrix...\n");
+
+    //loop through procs and set one to proc of interest (poi)
+    for (int proc=0; proc < comm->nprocs; proc++) {
+        // obtain nlocal of poi
+        // atoms 0->nlocal-1 belong to poi
+        plocal = atom->nlocal;
+        //broadcast poi's nlocal to other procs
+        MPI_Bcast(&plocal, 1, MPI_INT, proc, MPI_COMM_WORLD);
+        //every proc loops over poi's nlocal count
+        for (int i = 0; i < plocal; i++){
+            //ask if atom belongs to group
+            if (me == proc && mask[i] && groupbit)
+                group_flag = 1;
+            //broadcast answer to other procs
+            MPI_Bcast(&group_flag, 1, MPI_INT, proc, MPI_COMM_WORLD);
+            //only evaluate if atom did belong to group
+            if (group_flag) {
+                //obtain mass of aoi
+                if (me == proc)
+                {
+                    if (atom->rmass_flag == 1) {
+                        moaoi = m[i];
+                    }
+                    else
+                        moaoi = m[type[i]];
+                }
+                //broadcast mass of aoi
+                MPI_Bcast(&moaoi, 1, MPI_INT, proc, MPI_COMM_WORLD);
+                //obtain global id of local atom of interest (aoi)
+                if (me == proc) id = aid[i];
+                //broadcast global id of aoi to other procs
+                MPI_Bcast(&id, 1, MPI_INT, proc, MPI_COMM_WORLD);
+                //loop over x, y, z
+                for (int alpha = 0; alpha < 3; alpha++) {
+                    //move aoi
+                    if (me == proc) x[i][alpha] += del;
```

Comment:
Note, that you have to move not only the atom of interest on the processor that owns it, but also all its ghosts, either on the local processor or on remote processors.